### DISCUSSION TITLE
Circular Polarized source

### DISCUSSION THREAD

**Discussion by estebvac** [NanoComp/meep#2220] [Q&A] [answered]
Dear all, I know this is a quite old topic, but apart of the following description in scheme, I could not find any detailed and simple way to implement a Circular polarized source with the python API. 

https://meep-discuss.ab-initio.mit.narkive.com/Ulo0vcBL/creating-a-circular-polarized-source

Kindly,
Esteban

**Comment by smartalecH** [comment, marked as answer]
Although your linked response is for scheme, you can do the same thing using the python API. 

Namely, place two (orthogonal) linearly polarized, colocated sources that are 90 degrees out of phase. This is easy to do using the `amplitude` parameter of all `Source` objects, which simply multiplies the source by a constant (arbitrarily complex) scale factor. In fact you can produce any elliptical polarization this way, simply by changing the phase angle offset and relative amplitudes of each basis (ie meep source object). 

https://meep.readthedocs.io/en/latest/Python_User_Interface/#source

You can see how one does something similar in cylindrical coordinates:

https://meep.readthedocs.io/en/latest/FAQ/#how-do-i-create-a-circularly-polarized-planewave-source-in-cylindrical-coordinates

**Reply by estebvac** [reply]
Thanks for your answer, I followed the example you suggested, but I did not fund how to find the phase of the source in the API. Is the following script correct, or how should it be modified: 
``` python
sources = [mp.Source(mp.ContinuousSource(frequency=0.15),
                     component=mp.Ez,
                     center=mp.Vector3(0,0)),
           mp.Source(mp.ContinuousSource(frequency=0.15),
                     component=mp.Ex,
                     center=mp.Vector3(0,0),
                     amplitude=-1j)
          ]
```

**Reply by smartalecH** [reply]
> but I did not fund how to find the phase of the signal in the API

What signal? Phase relative to what? What exactly are you trying to do?

If you're just trying to verify that you're setting things up correctly, you can plot the fields for both components and see that they are indeed out of phase. 

Your code looks fine.

**Reply by estebvac** [reply]
In fact, I am just trying to verify if the configuration for a circular polarizer is correct. I followed your suggestion, plotting the response of each component of the source as follows:

```Python
import meep as mp 
import numpy as np
import matplotlib.pyplot as plt
import h5py


s = 5

geometry = [mp.Block(mp.Vector3(mp.inf,1,mp.inf),
                     center=mp.Vector3(),
                     material=mp.Medium(epsilon=12))]


sources =  [[ mp.Source(mp.ContinuousSource(frequency=1),
                     component=mp.Ex,
                     center=mp.Vector3(-7,0),
                     size=mp.Vector3(0,s,s),
                    ),],
            [
             mp.Source(mp.ContinuousSource(frequency=1),
                     component=mp.Ez,
                     center=mp.Vector3(-7,0),
                     size=mp.Vector3(0,s,s),
                     amplitude=np.exp( (np.pi/2) * (0+1j) ),
                    ),
            ]]

for ix, source in enumerate(sources):
    pml_layers = [mp.PML(1.0)]    
    cell_size = mp.Vector3(16,5,5)
    resolution = 10
    
    sim = mp.Simulation(cell_size=cell_size,
                        boundary_layers=pml_layers,
                        resolution=resolution,
                        sources=source,
                        )
    

    sim.reset_meep()
    sim.run(until=50)
    
    plt.figure(figsize=(20,10))
    plane = [16,5,0]
    output_plane = mp.Volume(center=mp.Vector3(), size=mp.Vector3(plane[0], plane[1], plane[2]))
    plt.subplot(3,2,1)
    sim.plot2D(output_plane=output_plane,fields=mp.Ex)
    plt.title(f'Ex, source{ix}, plane XY')
    plt.subplot(3,2,2)
    sim.plot2D(output_plane=output_plane,fields=mp.Ez)
    plt.title(f'Ez, source{ix}, plane XY')
    
    plane = [16,0,5]
    output_plane = mp.Volume(center=mp.Vector3(), size=mp.Vector3(plane[0], plane[1], plane[2]))
    
    plt.subplot(3,2,3)
    sim.plot2D(output_plane=output_plane,fields=mp.Ex)
    plt.title(f'Ex, source{ix}, plane XY')
    plt.subplot(3,2,4)
    sim.plot2D(output_plane=output_plane,fields=mp.Ez)
    plt.title(f'Ez, source{ix}, plane XY')
    
    plane = [0,5,5]
    output_plane = mp.Volume(center=mp.Vector3(), size=mp.Vector3(plane[0], plane[1], plane[2]))
    
    plt.subplot(3,2,5)
    sim.plot2D(output_plane=output_plane,fields=mp.Ex)
    plt.title(f'Ex, source{ix}, plane YZ')
    plt.subplot(3,2,6)
    sim.plot2D(output_plane=output_plane,fields=mp.Ez)
    plt.title(f'Ez, source{ix}, plane YZ')
```
The field produced by the first source has this shape : 
![image](https://user-images.githubusercontent.com/21016728/188507589-32c7674a-e3a1-47e6-9934-57b363892c4a.png)
The field produced by the second is: 
![image](https://user-images.githubusercontent.com/21016728/188507625-9679f15a-f5bc-4927-bdbf-a3e71fd80a44.png)

I just need to double-check if the configuration for a circular polarized source is correct. thanks in advance, and sorry for the long text.
